Make empty bb and bw files for zero analyzedReads

cadley-nyulangone opened the pull request:
When analyzedReads==0, make empty bb and bw files so that the blank tracks can appear in the browser.

Special non-empty bw files need to be constructed, as bedGraphToBigWig fails for zero reads.

This runs in my test rig, but can we give this a test with the main code?  My test sample is:
Bl6_Cast_dPiga_dIgf2_LP300_D5_dSox2_B4-BS08269A.cegsvectors_pSpCas9GFP

mattmaurano commented on dnase/callsnpsMerge.sh:
```diff
@@ -29,6 +29,7 @@ analysisType=${2}
 sampleOutdir=${3}
 sampleAnnotation=${4}
 src=${5}
+analyzedReads=${6}
```
can we check whether the bedgraph files are empty rather than adding a new argument?

mattmaurano commented on dnase/callsnpsMerge.sh:
```diff
@@ -77,12 +78,17 @@ rm -f ${covfiles}
-#Fix problems with reads running off end of chromosome
-bedClip $TMPDIR/${name}.${mappedgenome}.coverage.bedGraph ${chromsizes} $TMPDIR/${name}.${mappedgenome}.coverage.clipped.bedGraph
+if [[ "${analyzedReads}" == 0 ]]; then
+    bedtools genomecov -bg -i <(echo -e "fakeChrom\t0\t1\tfakeReadID\t0\t+") -g <(echo -e "fakeChrom\t10") > $TMPDIR/${name}.${mappedgenome}.coverage.clipped.bedGraph
```
how about just creating the dummy bedGraph file in the if statement, so the bedGraphToBigWig line is common to both?

mattmaurano commented on dnase/analysis.sh:
```diff
@@ -15,7 +15,6 @@ BS=${4}
 sampleAnnotation=${5}
 src=${6}
 
-
```
stray change

mattmaurano commented:
John, instead of making a dummy bedGraph and chromsizes file, how about taking the first entry in the real chromsizes and making a single-line dummy bedGraph with score=0. That will simplify the code and avoid inventing a fake chromosome. 0 is perfectly accurate in this case.

mattmaurano commented:
Also, this PR should remove the int(curSample['analyzed_reads'])>0 check in MakeTrackhub.py

cadley-nyulangone commented:
I implemented the score=0 idea.  Before I make the PR, I want to make sure
you are OK with how it appears.

The "coverage.bw" track in the middle of the screen below is the empty
one.  I made a bedgraph file with start=0, end=1, score=0, and converted
that to a bw file. Just displaying the first 5 bps, we get this picture:

[image: image.png]

You can see the horizontal line in position #1, which is the zero score for
the read.

When the bw view is compressed, we get this picture:

[image: image.png]

The grey bar in position #1 is the browser's attempt to show that something
is there.

Zooming out to the maximum view, we get this:

[image: image.png]

There is a very small vertical line on the far left, just to the right of
the bw label.   That's a very short version of the grey bar we saw when
only 5 bps were displayed.

When the bw view is expanded again, there is a very, very small dot on the
left, just to the right of the "0".  That's a very, very small version of
the horizontal line in the first screen shot above.

[image: image.png]

Should I use this in the PR ?

John


On Tue, Dec 1, 2020 at 4:11 PM mattmaurano <notifications@github.com> wrote:

> Also, this PR should remove the int(curSample['analyzed_reads'])>0 check
> in MakeTrackhub.py
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mauranolab/mapping/pull/240#issuecomment-736822964>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKY4NJRQVDJ6WGABI6QNMJDSSVLZHANCNFSM4UIECZHA>
> .
>

cadley-nyulangone commented:
Also added talOcc4 to the CEGS_genomes file.   This should have been in git long ago.

mattmaurano commented on dnase/trackhub/assets/CEGS_genomes:
```diff
@@ -1,4 +1,5 @@
 hg38
 mm10
 rn6
+talOcc4
```
Commit the CEGS_genomes as a separate PR (I will squash and merge this one)

mattmaurano commented on dnase/callsnpsMerge.sh:
```diff
@@ -96,8 +102,14 @@ rm -f ${allreadscovfiles}
+if [ ! -s "$TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph" ]; then
+    # The coverage.allreads.bedGraph file is empty
+    bedgraph_chrom=$(head -n 1 ${chromsizes} | cut -f1)
+    echo -e "${bedgraph_chrom}\t0\t0\t0" > $TMPDIR/${name}.${mappedgenome}.coverage.allreads.clipped.bedGraph
```
write to to TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph so you can move the bedclip out of the else block. It doesn't need to be clipped, but this makes the logic clearer

cadley-nyulangone commented:
When I take bedclip out of the else block, it tries to process the
artificial coverage.bedgraph file.  When that file is constructed like
this:  [chrom 0 0 0] , then bedlcip generates an
empty .coverage.clipped.bedGraph file.
This empty file causes bedGraphToBigWig to generate the needLargeMem error.

When the coverage.bedgraph file looks like this:  [chrom 0 1 0], all is
well, but it generates those stray marks in the browser screen.

I think I need to leave bedclip in the else block.



On Wed, Dec 2, 2020 at 1:12 PM mattmaurano <notifications@github.com> wrote:

> *@mattmaurano* requested changes on this pull request.
> ------------------------------
>
> In dnase/trackhub/assets/CEGS_genomes
> <https://github.com/mauranolab/mapping/pull/240#discussion_r534377612>:
>
> > @@ -1,4 +1,5 @@
>  hg38
>  mm10
>  rn6
> +talOcc4
>
> Commit the CEGS_genomes as a separate PR (I will squash and merge this one)
> ------------------------------
>
> In dnase/callsnpsMerge.sh
> <https://github.com/mauranolab/mapping/pull/240#discussion_r534378522>:
>
> > @@ -96,8 +102,14 @@ rm -f ${allreadscovfiles}
>
>  unstarch ${sampleOutdir}/${name}.${mappedgenome}.coverage.allreads.starch | cut -f1-3,5 > $TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph
>
> -#Fix problems with reads running off end of chromosome
> -bedClip $TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph ${chromsizes} $TMPDIR/${name}.${mappedgenome}.coverage.allreads.clipped.bedGraph
> +if [ ! -s "$TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph" ]; then
> +    # The coverage.allreads.bedGraph file is empty
> +    bedgraph_chrom=$(head -n 1 ${chromsizes} | cut -f1)
> +    echo -e "${bedgraph_chrom}\t0\t0\t0" > $TMPDIR/${name}.${mappedgenome}.coverage.allreads.clipped.bedGraph
>
> write to to TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph so
> you can move the bedclip out of the else block. It doesn't need to be
> clipped, but this makes the logic clearer
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mauranolab/mapping/pull/240#pullrequestreview-543108124>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKY4NJWG3NQN45S2C73NDUDSSZ7P7ANCNFSM4UIECZHA>
> .
>

cadley-nyulangone commented:
I was just thinking about the change to MakeTrackhub.py

Once the check for zero reads is removed, it will try to process old
flowcells for which there might not be bb and bw files.

On Wed, Dec 2, 2020 at 1:45 PM John Cadley <cadley.nyulangone@gmail.com>
wrote:

> When I take bedclip out of the else block, it tries to process the
> artificial coverage.bedgraph file.  When that file is constructed like
> this:  [chrom 0 0 0] , then bedlcip generates an
> empty .coverage.clipped.bedGraph file.
> This empty file causes bedGraphToBigWig to generate the needLargeMem error.
>
> When the coverage.bedgraph file looks like this:  [chrom 0 1 0], all is
> well, but it generates those stray marks in the browser screen.
>
> I think I need to leave bedclip in the else block.
>
>
>
> On Wed, Dec 2, 2020 at 1:12 PM mattmaurano <notifications@github.com>
> wrote:
>
>> *@mattmaurano* requested changes on this pull request.
>> ------------------------------
>>
>> In dnase/trackhub/assets/CEGS_genomes
>> <https://github.com/mauranolab/mapping/pull/240#discussion_r534377612>:
>>
>> > @@ -1,4 +1,5 @@
>>  hg38
>>  mm10
>>  rn6
>> +talOcc4
>>
>> Commit the CEGS_genomes as a separate PR (I will squash and merge this
>> one)
>> ------------------------------
>>
>> In dnase/callsnpsMerge.sh
>> <https://github.com/mauranolab/mapping/pull/240#discussion_r534378522>:
>>
>> > @@ -96,8 +102,14 @@ rm -f ${allreadscovfiles}
>>
>>  unstarch ${sampleOutdir}/${name}.${mappedgenome}.coverage.allreads.starch | cut -f1-3,5 > $TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph
>>
>> -#Fix problems with reads running off end of chromosome
>> -bedClip $TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph ${chromsizes} $TMPDIR/${name}.${mappedgenome}.coverage.allreads.clipped.bedGraph
>> +if [ ! -s "$TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph" ]; then
>> +    # The coverage.allreads.bedGraph file is empty
>> +    bedgraph_chrom=$(head -n 1 ${chromsizes} | cut -f1)
>> +    echo -e "${bedgraph_chrom}\t0\t0\t0" > $TMPDIR/${name}.${mappedgenome}.coverage.allreads.clipped.bedGraph
>>
>> write to to TMPDIR/${name}.${mappedgenome}.coverage.allreads.bedGraph so
>> you can move the bedclip out of the else block. It doesn't need to be
>> clipped, but this makes the logic clearer
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/mauranolab/mapping/pull/240#pullrequestreview-543108124>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AKY4NJWG3NQN45S2C73NDUDSSZ7P7ANCNFSM4UIECZHA>
>> .
>>
>

mattmaurano commented:
John, can you commit the CEGS_genomes as a separate PR and remove it from this one? I will squash and merge this one.

cadley-nyulangone commented:
Did it.

On Thu, Dec 3, 2020 at 8:45 PM mattmaurano <notifications@github.com> wrote:

> John, can you commit the CEGS_genomes as a separate PR and remove it from
> this one? I will squash and merge this one.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mauranolab/mapping/pull/240#issuecomment-738499834>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKY4NJRGUMJ3OO373LFIQS3STA5LPANCNFSM4UIECZHA>
> .
>
